### PR TITLE
[Blockchain]: Move Ethereum implementation to Rust

### PR DESCRIPTION
**Pull request by satoshiotomakan**
## Description

The first step of moving Ethereum blockchain from C++ to Rust.

## TODO

* Redesign Ethereum Message Signing and Ethereum ABI interfaces
* Implement `tw_ethereum::Entry::sign_json` - we need to replace [quick-protobuf](https://crates.io/crates/quick-protobuf) with [rust-protobuf](https://crates.io/crates/protobuf3) as it's the only crate that supports `Protobuf -> JSON` conversion

## Types of changes

Important changes that have been made in existing logic:
* VeChain and Theta no longer support JSON signing. Initially, the chains inherited `Ethereum::Entry`, but `Entry::signJSON` was not overridden with the chain specific implementation.
* `UserOperation` signed transaction (JSON representation) now has `sender` normalized with EIP55 checksum.
* `Metis` blockchain had an incorrect `coinId` - it was different in `registry.json` and `TWCoinType.h`. Changed from 10001088 to 1001088.
* Moved `Ronin` chain into Rust to be able to remove `Ethereum/Signer` and `Ethereum/Transaction` C++ impl

## How to test

Run Rust, C++, Swift, Kotlin tests

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

### CODE REVIEW

File: rust/tw_evm/src/abi/prebuild/erc1155.abi.json
Comment (lamafab) [inline rust/tw_evm/src/abi/prebuild/erc1155.abi.json:1]: I'd maybe put those json files into a separate directly (`json/` or so), but that's just a nitpick.

**Review by lamafab — APPROVED** [review]
Regarding https://github.com/trustwallet/wallet-core/pull/3295#issuecomment-1645108469

>  We already have `wallet_core_rs` crate that could contain all of the FFI implementations.

Yes, that works too.

> What do you think if I do that refactoring first in master branch and then merge into dev?

I'm not quite sure what you mean by this?

**Comment by satoshiotomakan** [comment]
I mean I would like to do this refactoring in `master` branch first
> What do you think if I do that refactoring first in master branch and then merge into dev?

> I'm not quite sure what you mean by this?

**Review by Milerius — COMMENTED** [review]
First review iteration 30files / 128, minor comments

File: registry.json [hunk 1]
```diff
@@ -3669,7 +3669,7 @@
   {
     "id": "metis",
     "name": "Metis",
-    "coinId": 10001088,
+    "coinId": 1001088,
```
[the remark this answers is not in the document]
Comment (satoshiotomakan) [reply registry.json:3672]: Explained here https://github.com/trustwallet/wallet-core/pull/3322#discussion_r1273323868

File: rust/tw_any_coin/src/ffi/tw_any_address.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+/// \param string address to validate.
+/// \param coin coin type of the address.
+/// \return bool indicating if the address is valid.
+#[no_mangle]
+pub unsafe extern "C" fn tw_any_address_is_valid(string: *mut TWString, coin: u32) -> bool {
+    let string = try_or_false!(TWString::from_ptr_as_ref(string));
+    let string = try_or_false!(string.as_str());
+
+    AnyAddress::is_valid(coin, string, None)
+}
```
[the remark this answers is not in the document]
Comment (satoshiotomakan) [reply rust/tw_any_coin/src/ffi/tw_any_address.rs:33]: Agree with that! Thank you

File: rust/tw_any_coin/src/ffi/tw_transaction_compiler.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        amount: try_str!(amount),
+        asset: try_str!(asset),
+        memo: try_str!(memo),
+        chain_id: try_str!(chain_id),
+    };
+
+    TransactionCompiler::build_input(coin, input)
+        .map(|output| TWData::from(output).into_ptr())
+        .unwrap_or_else(|_| std::ptr::null_mut())
+}
```
[the remark this answers is not in the document]
Comment (satoshiotomakan) [reply rust/tw_any_coin/src/ffi/tw_transaction_compiler.rs:59]: I'd be happy to get rid of this, let me check if it's used in any end-user app

[the remark this answers is not in the document]
Comment (satoshiotomakan) [reply rust/tw_any_coin/src/ffi/tw_transaction_compiler.rs]: DeFi wallet doesn't use `TWTransactionCompilerBuildInput` anymore. TW end-user apps don't use it, so I guess we can deprecate it? I'd suggest not removing the interface at all, but leave it implemented for Binance chain only. So EVM compatible chains will not longer implement that functionality
@Milerius what do you think?

[the remark this answers is not in the document]
Comment (Milerius) [reply rust/tw_any_coin/src/ffi/tw_transaction_compiler.rs]: make sense but then we don't rly need it in rust tbh

[the remark this answers is not in the document]
Comment (satoshiotomakan) [reply rust/tw_any_coin/src/ffi/tw_transaction_compiler.rs]: Yes yes, of course

File: rust/tw_evm/src/abi/prebuild/erc1155.abi.json
Comment (satoshiotomakan) [reply rust/tw_evm/src/abi/prebuild/erc1155.abi.json:1]: Done

File: rust/tw_ethereum/src/entry.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    #[inline]
+    fn json_signer(&self) -> Option<Self::JsonSigner> {
+        Some(EthJsonSigner::default())
+    }
+
+    #[inline]
+    fn plan_builder(&self) -> Option<Self::PlanBuilder> {
+        None
+    }
```
Comment (Milerius) [inline rust/tw_ethereum/src/entry.rs:92]: Would be better if coin entry have plan_builder that return None by default, and force chain that have a plan to implement it no?

File: rust/tw_ethereum/tests/signer.rs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+#[test]
+fn test_plan_not_supported() {
+    let input = Proto::SigningInput::default();
+    let input_data = serialize(&input).unwrap();
+    let maybe_plan = EthereumEntry
+        .plan(&EmptyCoinContext, &input_data)
+        .expect("!plan");
+    assert_eq!(maybe_plan, None, "Ethereum does not support plan()");
+}
```
Comment (Milerius) [inline rust/tw_ethereum/tests/signer.rs:36]: same, that's sad if we have to do it for every blockchain

File: rust/tw_evm/src/abi/mod.rs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+/// TODO refactor this by implementing `RlpEncode` for `U256` at the next iteration.
+pub fn convert_u256(num: U256) -> ethabi::Uint {
+    let bytes = num.to_big_endian().take();
+    ethabi::Uint::from_big_endian(&bytes)
+}
+
+/// TODO refactor this by implementing `RlpEncode` for `Address` at the next iteration.
+pub fn convert_address(addr: Address) -> ethabi::Address {
+    ethabi::Address::from(addr.bytes().take())
+}
```
Comment (Milerius) [inline rust/tw_evm/src/abi/mod.rs:41]: Let's open issue to not forget that

File: rust/tw_evm/src/abi/prebuild/resource/erc4337.simple_account.sol [hunk 7]
```diff
@@ -0,0 +1,139 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.12;
+
```
Comment (Milerius) [inline rust/tw_evm/src/abi/prebuild/resource/erc4337.simple_account.sol:3]: is this file needed to be pushed?

File: rust/tw_evm/src/address.rs [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+    /// Tries to parse an address from the string representation.
+    /// Returns `Ok(None)` if the given `s` string is empty.
+    #[inline]
+    fn from_str_optional(s: &str) -> AddressResult<Option<Self>> {
+        if s.is_empty() {
+            return Ok(None);
+        }
+
+        Self::from_str(s).map(Some)
+    }
```
Comment (Milerius) [inline rust/tw_evm/src/address.rs:27]: Is it possible to do something like `(!s.is_empty()).then(|| Self::from_str(s)).transpose()` ?

File: rust/tw_evm/src/address.rs [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    bytes: H160,
+}
+
+/// cbindgen:ignore
+impl Address {
+    pub const LEN: usize = 20;
+
+    /// Initializes an address with a `secp256k1` public key.
+    pub fn with_secp256k1_pubkey(pubkey: &secp256k1::PublicKey) -> Address {
+        const ADDRESS_HASH_RANGE: RangeFrom<usize> = 12..;
```
Comment (Milerius) [inline rust/tw_evm/src/address.rs:42]: what's 12 here? Let's use a constant?

File: rust/tw_evm/src/address.rs [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+
+        let pubkey_bytes = pubkey.uncompressed_without_prefix();
+        let hash = keccak256(pubkey_bytes.as_slice());
+        assert_eq!(hash.len(), 32);
```
Comment (Milerius) [inline rust/tw_evm/src/address.rs:46]: 32 seems to be required hash size, can we wrap it into a constant? Actually I'm quite sure that H160:: have something for constant size no?

File: rust/tw_evm/src/address.rs [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+                return a;
+            }
+            if UPPER_RANGE_1.contains(&h) || UPPER_RANGE_2.contains(&h) {
+                a.to_ascii_uppercase()
+            } else {
+                a.to_ascii_lowercase()
+            }
+        });
+
+        "0x".chars().chain(payload_chars).collect()
```
Comment (Milerius) [inline rust/tw_evm/src/address.rs:74]:
```rust
let payload_chars = addr_hex.chars().zip(addr_hash.chars()).map(|(a, h)| {
        if a.is_ascii_digit() {
            a
        } else if (h <= UPPER_BOUND_1) || (h >= LOWER_BOUND_2 && h <= UPPER_BOUND_2) {
            a.to_ascii_uppercase()
        } else {
            a.to_ascii_lowercase()
        }
    });

    ["0x", &payload_chars.collect::<String>()].concat()
    ```
    this way? let's move the constant at module level?

**Review by Milerius — COMMENTED** [review]
Second review iteration, minor comments

File: rust/tw_evm/src/abi/prebuild/resource/erc4337.simple_account.sol (hunk 7 above)
Comment (satoshiotomakan) [reply rust/tw_evm/src/abi/prebuild/resource/erc4337.simple_account.sol:3]: I think yes because I had to modify the original https://github.com/eth-infinitism/account-abstraction/blob/develop/contracts/samples/SimpleAccount.sol
So it will be easier to make any changes in the ERC4337 protocol later

File: rust/tw_evm/src/address.rs (hunk 8 above)
Comment (satoshiotomakan) [reply rust/tw_evm/src/address.rs:27]: I don't think so, @lamafab do you have any ideas?

File: rust/tw_evm/src/address.rs (hunk 9 above)
Comment (satoshiotomakan) [reply rust/tw_evm/src/address.rs:42]: This is a constant itself 😅 
I can add a comment on why we start from 12's byte

File: rust/tw_evm/src/address.rs (hunk 10 above)
Comment (satoshiotomakan) [reply rust/tw_evm/src/address.rs:46]: I dream to refactor `keccak256` and other hash functions to return a strict `H256`, `H512` etc. So this check will be no longer required. But I agree with you that better to use `H256::LEN` here

File: rust/tw_evm/src/abi/prebuild/resource/erc4337.simple_account.sol (hunk 7 above)
Comment (Milerius) [reply rust/tw_evm/src/abi/prebuild/resource/erc4337.simple_account.sol:3]: ok @rsrbk let's also review the change on this file please!

File: rust/tw_evm/src/address.rs (hunk 9 above)
Comment (Milerius) [reply rust/tw_evm/src/address.rs:42]: yes please, let's name the constant in a variable if possible

File: rust/tw_evm/src/address.rs (hunk 11 above)
Comment (satoshiotomakan) [reply rust/tw_evm/src/address.rs]: Done

File: rust/tw_evm/src/address.rs (hunk 9 above)
Comment (satoshiotomakan) [reply rust/tw_evm/src/address.rs]: Done

File: rust/tw_evm/src/abi/prebuild/resource/erc4337.simple_account.sol (hunk 7 above)
Comment (satoshiotomakan) [reply rust/tw_evm/src/abi/prebuild/resource/erc4337.simple_account.sol:3]: Btw, the original `SimpleAccount.sol` has been updated 4 days ago!
https://github.com/eth-infinitism/account-abstraction/commit/0a9d55ca73cd309f4ec1cda28284d2d039e552b9
I can remove it now

Comment (Milerius) [reply rust/tw_evm/src/abi/prebuild/resource/erc4337.simple_account.sol:3]: alright

Comment (satoshiotomakan) [reply rust/tw_evm/src/abi/prebuild/resource/erc4337.simple_account.sol]: Done

File: rust/tw_evm/src/abi/mod.rs (hunk 6 above)
Comment (satoshiotomakan) [reply rust/tw_evm/src/abi/mod.rs:41]: https://github.com/trustwallet/wallet-core/issues/3335

**Review by rsrbk — APPROVED** [review]
Checked 4337 stuff, looks good!

File: rust/tw_evm/src/abi/prebuild/erc1155.rs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+/// Solidity: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.9.2/contracts/token/ERC1155/IERC1155.sol
+const ERC1155_ABI: &str = include_str!("resource/erc1155.abi.json");
+
+lazy_static! {
+    static ref ERC1155: Contract = serde_json::from_str(ERC1155_ABI).unwrap();
+}
+
+pub struct Erc1155;
+
+impl Erc1155 {
```
Comment (a6-dou) [inline rust/tw_evm/src/abi/prebuild/erc1155.rs:24]: not sure if you are planning to add full support but if you do then we need to support:
- safeBatchTransferFrom
- balanceOf
- balanceOfBatch
- setApprovalForAll
- isApprovedForAll

File: rust/tw_evm/src/abi/prebuild/erc721.rs [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+lazy_static! {
+    static ref ERC721: Contract = serde_json::from_str(ERC721_ABI).unwrap();
+}
+
+pub struct Erc721;
+
+impl Erc721 {
+    pub fn encode_transfer_from(from: Address, to: Address, token_id: U256) -> AbiResult<Data> {
+        let func = ERC721.function("transferFrom")?;
```
Comment (a6-dou) [inline rust/tw_evm/src/abi/prebuild/erc721.rs:26]: most of the time `safeTransferFrom` is used instead of `transferFrom`,
also `setApprovalForAll ` and `isApprovedForAll ` functions are used more often than other functions.
ideally adding support for the whole ERC makes more sense, but if you want to add minimum abstract support, i would suggest adding the 3 functions i mentioned above

File: rust/tw_evm/src/modules/tx_builder.rs [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+            },
+            Tx::erc20_approve(ref erc20_approve) => {
+                let spender = Self::parse_address(&erc20_approve.spender)?;
+                let token_amount = U256::from_big_endian_slice(&erc20_approve.amount)?;
+                let contract_address = Self::parse_address(&input.to_address)?;
+
+                let payload = Erc20::approve(spender, token_amount)?;
+                (U256::zero(), payload, Some(contract_address))
+            },
+            Tx::erc721_transfer(ref erc721_transfer) => {
```
[the remark this answers is not in the document]
Comment (satoshiotomakan) [reply rust/tw_evm/src/modules/tx_builder.rs:62]: As I said [here](https://github.com/trustwallet/wallet-core/pull/3322#issuecomment-1653716099), this PR aims to move existing functionality from C++ to Rust.
Here is the function calls that we support right now:
https://github.com/trustwallet/wallet-core/blob/master/src/proto/Ethereum.proto#L93-L99

**Review by Milerius — APPROVED** [review]

File: rust/tw_evm/src/abi/prebuild/erc1155.rs (hunk 12 above)
Comment (satoshiotomakan) [reply rust/tw_evm/src/abi/prebuild/erc1155.rs:24]: Thank you, will put it to the list!
